### PR TITLE
Added an ImplicitProjectPropertiesStore to sit at the UnconfiguredProject level

### LINKED ISSUE #392: "Unrecognized Guid format" when Add Class
1. Right-click on folder -> Add -> Add Class

image

### PR DESCRIPTION
**Pull request by 333fred**
The ImplicitProjectPropertiesProvider sits at the configured project level. However, the project guid is set once per UnconfiguredProject. When a different configuration was used from the initial default configuration, this guid wasn't being retained. This introduced a store that sits at the UnconfiguredProject level that the ImplicitProjectPropertiesProvider imports, sharing properties like the project guid among all configured projects. This fixes #392 when switching build configurations, and should fully fix it.

### CODE REVIEW

**Comment by 333fred** [comment]
Tagging @davkean @dotnet/project-system for review.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ImplicitProjectPropertiesStore.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    /// be shared by all configured projects, each of which will get its own <see cref="ImplicitProjectPropertiesProvider"/>.
+    /// </summary>
+    [Export(typeof(ImplicitProjectPropertiesStore<,>))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class ImplicitProjectPropertiesStore<T1, T2> : ConcurrentDictionary<T1, T2>
+    {
+        // We import UnconfiguredProject here to ensure that we're loaded into the UnconfiguredProject scope.
+        // However, we don't need the project for anything.
+        [ImportingConstructor]
+        public ImplicitProjectPropertiesStore(UnconfiguredProject project) { }
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ImplicitProjectPropertiesStore.cs]: Instead you can add `[ProjectSystemContract(ProjectSystemContractScope.UnconfiguredProject, ProjectSystemContractProvider.Host)]` attribute to the class which will export this class in the `UnconfiguredScope`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ImplicitProjectPropertiesProvider.cs [hunk 2]
```diff
@@ -25,19 +25,22 @@ namespace Microsoft.VisualStudio.ProjectSystem.Properties
...
 
         [ImportingConstructor]
         public ImplicitProjectPropertiesProvider(
             [Import(ContractNames.ProjectPropertyProviders.ProjectFile)] IProjectPropertiesProvider provider,
             [Import(ContractNames.ProjectPropertyProviders.ProjectFile)] IProjectInstancePropertiesProvider instanceProvider,
+            ImplicitProjectPropertiesStore<string, string> propertyStore,
             UnconfiguredProject unconfiguredProject)
             : base(provider, instanceProvider, unconfiguredProject)
         {
+            Requires.NotNull(propertyStore, nameof(propertyStore));
```
Comment (basoundr) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ImplicitProjectPropertiesProvider.cs]: As @BillHiebert mentioned in other PR, the MEF imported components are implicitly checked for null. Hence this null check is not necessary

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ImplicitProjectPropertiesStore.cs (hunk 1 above)
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ImplicitProjectPropertiesStore.cs]: I didn't know this, thanks.

**Comment by 333fred** [comment]
CI seems to be having issues publishing the builds. It passes:

* http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-project-system/job/master/job/windows_debug_prtest/371/console
* http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-project-system/job/master/job/windows_release_prtest/378/console

**Comment by basoundr** [comment]
:+1:

**Comment by 333fred** [comment]
Thanks @basoundr

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ImplicitProjectPropertiesStore.cs (hunk 1 above)
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ImplicitProjectPropertiesStore.cs]: Just to clarify this - this is not how this works. This attribute does not control the scope in which you end up in.

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/ImplicitProjectPropertiesStore.cs]: See https://github.com/Microsoft/VSProjectSystem/blob/master/doc/overview/scopes.md for more information.

**Pull request merged**